crates/transaction-pool/src/pool/pending.rs

```diff
@@ -325,10 +352,126 @@ impl<T: TransactionOrdering> PendingPool<T> {
         id
     }
 
-    /// Removes the worst transaction from this pool.
-    pub(crate) fn pop_worst(&mut self) -> Option<Arc<ValidPoolTransaction<T::Transaction>>> {
-        let worst = self.all.iter().next().map(|tx| *tx.transaction.id())?;
-        self.remove_transaction(&worst)
+    /// Traverses the pool, starting at the highest nonce set, returning the transactions which
+    /// would put the pool under the specified limits.
+    ///
+    /// If the `local` flag is unset, then only non-local transactions will be removed.
+    pub(crate) fn transactions_to_remove(
+        &self,
+        limit: &SubPoolLimit,
+        local: bool,
+    ) -> Vec<TransactionId> {
+        // this serves as a termination condition for the loop - it represents the number of
+        // _valid_ unique senders that might have descendants in the pool.
+        //
+        // If `local` is false, a value of zero means that there are no non-local txs in the pool
+        // that can be removed.
+        //
+        // If `local` is true, a value of zero means that there are no txs in the pool that can be
+        // removed.
+        let mut unique_senders = self.highest_nonces.len();
+
+        // transactions to remove
+        let original_length = self.len();
+        let mut removed = Vec::new();
+
+        // track total `size` of transactions to remove
+        let original_size = self.size();
+        let mut total_size = 0;
+
+        // starting loop to populate `removed`
+        for tx in self.highest_nonces.iter() {
+            // return early if the pool is under limits
+            if original_size - total_size <= limit.max_size &&
+                original_length - removed.len() <= limit.max_txs
+            {
+                return removed
+            }
+
+            if !local && tx.transaction.is_local() {
+                unique_senders -= 1;
+                continue
+            }
+
+            total_size += tx.transaction.size();
+            removed.push(*tx.transaction.id());
+        }
+
+        // index of first tx to check
+        let mut removed_idx = 0;
+        'outer: loop {
```

Comment:
not immediately obvious why we need this and this looks a bit complex.

I guess we need this because removing 1 tx per unique sender might not be sufficient? But the highest nonce is only updated once we remove the transactions?

I think this can be simplified by putting L372-L397 in a loop that actually removes the recorded ids of `removed` so we don't need the loop below?

If the function takes a `&mut Vec` then the transactions can be pushed directly to this, for both (locals =true|false) calls



Under real circumstances this will only evict 1 to a few transactions so having a tmp buffer for the `removed` ids isn't too bad.